jmclient/jmclient/fidelity_bond.py

```diff
@@ -0,0 +1,136 @@
+        self.cert_privkey = cert_privkey
+        self.cert_pubkey = cert_pubkey
+        self.cert_signature = cert_signature
+
+    def create_proof(self, maker_nick, taker_nick):
+        return FidelityBondProof(
+            maker_nick, taker_nick, self.cert_pubkey, self.cert_expiry,
+            self.cert_signature, self.utxo, self.utxo_pubkey, self.locktime)
+
+    def serialize(self):
+        return json.dumps([
+            self.utxo,
+            self.utxo_pubkey,
+            self.locktime,
+            self.cert_expiry,
+            self.cert_privkey,
+            self.cert_pubkey,
+            self.cert_signature,
+        ])
+
+    @classmethod
+    def deserialize(cls, data):
+        return cls(*json.loads(data))
+
+
+class FidelityBondProof:
+    # nick_sig + cert_sig + cert_pubkey + cert_expiry + utxo_pubkey + txid + vout + timelock
+    # 72       + 72       + 33          + 2           + 33          + 32   + 4    + 4 = 252 bytes
+    SER_STUCT_FMT = '<72s72s33sH33s32sII'
+
+    def __init__(self, maker_nick, taker_nick, cert_pub, cert_expiry,
+                 cert_sig, utxo, utxo_pub, locktime):
+        assert isinstance(maker_nick, str)
+        assert isinstance(taker_nick, str)
+        assert isinstance(cert_pub, bytes)
+        assert isinstance(cert_sig, bytes)
+        assert isinstance(utxo_pub, bytes)
+        assert isinstance(locktime, int)
+        assert_is_utxo(utxo)
+        self.maker_nick = maker_nick
+        self.taker_nick = taker_nick
+        self.cert_pub = cert_pub
+        self.cert_expiry = cert_expiry
+        self.cert_sig = cert_sig
+        self.utxo = utxo
+        self.utxo_pub = utxo_pub
+        self.locktime = locktime
+
+    @property
+    def nick_msg(self):
+        return (self.taker_nick + '|' + self.maker_nick).encode('ascii')
+
+    def create_proof_msg(self, cert_priv):
+        nick_sig = ecdsa_sign(self.nick_msg, cert_priv)
+        # FIXME: remove stupid base64
+        nick_sig = base64.b64decode(nick_sig)
+        return self._serialize_proof_msg(nick_sig)
+
+    def _serialize_proof_msg(self, msg_signature):
+        msg_signature = msg_signature.rjust(72, b'\xff')
+        cert_sig = self.cert_sig.rjust(72, b'\xff')
+        fidelity_bond_data = struct.pack(
+            self.SER_STUCT_FMT,
+            msg_signature,
+            cert_sig,
+            self.cert_pub,
+            self.cert_expiry,
+            self.utxo_pub,
+            self.utxo[0],
+            self.utxo[1],
+            self.locktime
+        )
+        return base64.b64encode(fidelity_bond_data).decode('ascii')
+
+    @staticmethod
+    def _verify_signature(message, signature, pubkey):
+        # FIXME: remove stupid base64
+        return ecdsa_verify(message, base64.b64encode(signature), pubkey)
+
+    @classmethod
+    def parse_and_verify_proof_msg(cls, maker_nick, taker_nick, data):
+        if not fidelity_bond_sanity_check.fidelity_bond_sanity_check(data):
+            return None
```

Comment:
imo it's a code smell to return different data instead of raising an exception if something exceptional happens. If something happens that actually breaks the desired flow and forces error handling (opposed to falling back to some default value or similar, where returning `None` would not indicate a flow break). Having `except` handlers in calling code makes it much more transparent to a reader what exactly is happening here instead of checking for and handling some weird return value like `False` or `None`, only to divert to aborting the program flow. This also allows the code to cleanly provide error messages (in the form of an exception message). I know there are lots of occurrences of that pattern in joinmarket, mostly commonly in the form of a tuple `(success, return_value, …)` but I think it should be avoided if possible.

Reply:
Yeah that's probably a good call, this came from me - I think I was influenced by working a lot with the lnd codebase at the time, where that pattern is some kind of de facto standard. There are probably some subtleties and gray areas here, though.

Reply:
I used this pattern because it allows more cleanly using list comprehension. Especially in `ob-watcher.py` a few times I wrote lines something like `parsed_bonds = [FidelityBondProof.parse_and_verify_proof_msg(fb["counterparty"], fb["takernick"], fb["proof"]) for fb in fbonds]`. If some of the data is invalid we want the loop to carry on because the orderbook watcher will display a zero in the html table, not to break out of the loop with an exception.

Please suggest an alternative way of doing it which uses exceptions? I can think of using a for-loop where the loop body catches an exception and then builds up a list. IDK it seems a bit less clean to me but I'll do it if you reckon its important.

This function is parsing data which is received directly from another peer, that peer could be anyone and could be some spammer or DOSer, so it seems pointless to display an error message anywhere(?)

Reply:
I see your point about list comprehensions. I think using list comprehensions for more complex processing can make code very hard to read. `get_fidelity_bond_data` in `ob-watcher.py` is a good example. You're creating a lot of lists only to use the result in the next list comprehension. A `for` loop would be much cleaner and easier to read. As a side note, the code looks awfully similar to the old code of `calculate_fidelity_bond_values()`. copy & paste?

While showing the user an error about what exactly went wrong there may not be useful in this particular case, raising proper exceptions can aid well in debugging and testing. For example there are a lot of tests that have an `assert retval[0]` and it can be a little annoying to figure out what exactly went wrong there if the assert failed. Raising an exception instead would immediately tell you.

Reply:
I see your point about list comprehensions. I think list comprehensions should not be used for complex logic and `get_fidelity_bond_data()`  in `ob-watcher.py` is a good example for that. You're creating a lot of lists only to iterate over them in the next list comprehension. A `for` loop would be much cleaner and easier to read. As a side note, the code looks awfully similar to the old code in `calculate_fidelity_bond_values()`. Copy & paste?

While in this particular case it may not be useful to tell the user what went wrong (though debug logging could be useful in many cases), raising proper exceptions can help a lot with figuring out what is going wrong, especially if you don't expect a fail. There are for example a lot of tests with an `assert retval[0]`, which can be annoying to figure out if they fail. An exception telling so would be much more helpful.

Reply:
OK you're right